### PR TITLE
Feature: LookupRangeCheck trait and minor improvements

### PR DESCRIPTION
**Pull request by YaoJGalteland**
### Key Changes from zcash/halo2 v0.3.0

1. Introduced a `LookupRangeCheck` trait that provides common methods for a lookup range check.
2. Extended the capabilities of the `EccConfig` (EccChip, SinsemillaConfig, SinsemillaChip, MerkleConfig, MerkleChip) to include additional functionalities provided by the Lookup type
3. Splitted big functions (SinsemillaChip::configure, hash_message) into smaller building blocks for better reuseability. 
4. Added the `EccPointQ` enum to handle either type of point for `Q`.
5. Introduced `MerkleSinsemillaInstructions` trait to extend the capabilities of `SinsemillaInstructions`, for specific cryptographic operations involving generic lookup.
6. Added `test_utils.rs` to implement functions used for circuit tests.
7. Added tests for ecc chip, merkle chip, sinsemilla chip, lookup range check and short range check to test that the pinned verification key (representing the constraint system) does not change.
8. Added tests for ecc chip, merkle chip, sinsemilla chip, lookup range check and short range check against stored proofs.
9. The test files generated in the above two tests are stored in the `tests/circuit_data` folder.
10. Added default generic arguments for PallasLookupRC to improve backward compatibility. This change adds default generic arguments for types implementing the PallasLookupRC trait, assigning them to `PallasLookupRCConfig` by default. This adjustment allows existing code to function without requiring users to explicitly specify PallasLookupRCConfig, thus minimizing breaking changes. Users will still need to bring the PallasLookupRC trait into scope using 'use ... PallasLookupRC'.

### CODE REVIEW

File: halo2_gadgets/src/ecc.rs [hunk 1]
```diff
@@ -901,6 +911,57 @@ pub(crate) mod tests {
...
+        if std::env::var_os("CIRCUIT_TEST_GENERATE_NEW_PROOF").is_some() {
+            let create_proof = || -> std::io::Result<()> {
+                let proof = Proof::create(&vk, &params, circuit).unwrap();
+                assert!(proof.verify(&vk, &params).is_ok());
+
+                let file = std::fs::File::create("src/circuit_proof_test_case_ecc.bin")?;
+                write_test_case(file, &proof)
+            };
+            create_proof().expect("should be able to write new proof");
+        }
```
Comment (PaulLaux) [inline halo2_gadgets/src/ecc.rs:954]: encapsulate into `conditionally_save_circuit_to_disk(vk,file_name)`

(everywhere)

File: halo2_gadgets/src/ecc.rs [hunk 2]
```diff
@@ -901,6 +911,57 @@ pub(crate) mod tests {
...
+
+        // Read the old proof into 'proof'
```
Comment (PaulLaux) [inline halo2_gadgets/src/ecc.rs:956]: ``// read proof from disk``

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 3]
```diff
@@ -77,107 +85,35 @@ impl<F: PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K> {
...
+        region: &mut Region<'_, F>,
+        element: AssignedCell<F, F>,
+        num_bits: usize,
+    ) -> Result<(), Error>;
 
     /// Range check on an existing cell that is copied into this helper.
     ///
     /// Returns an error if `element` is not in a column that was passed to
     /// [`ConstraintSystem::enable_equality`] during circuit configuration.
-    pub fn copy_check(
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:180]: why `pub` removed ? (all)

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 4]
```diff
@@ -77,107 +85,35 @@ impl<F: PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K> {
     /// # Side-effects
     ///
     /// Both the `running_sum` and `constants` columns will be equality-enabled.
-    pub fn configure(
+    fn configure(
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:88]: the documentation should be part of the trait or the implementation? 
I think implementation.

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 5]
```diff
@@ -493,11 +557,54 @@ mod tests {
...
+                let proof = {
+                    let test_case_bytes =
+                        fs::read("src/utilities/circuit_proof_test_case_lookup_range_check.bin")
+                            .unwrap();
+                    read_test_case(&test_case_bytes[..]).expect("proof must be valid")
+                };
+
+                // Verify the old proof with the new vk
+                assert!(proof.verify(&vk, &params).is_ok());
+            }
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:601]: as mentioned above

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 6]
```diff
@@ -542,34 +649,131 @@ mod tests {
...
             let prover = MockProver::<pallas::Base>::run(11, &circuit, vec![]).unwrap();
             assert_eq!(prover.verify(), Ok(()));
+
+            // generate vk
+            let vk = plonk::keygen_vk(&params, &circuit).unwrap();
+
+            // Test that the pinned verification key (representing the circuit)
+            // is as expected. Which indicates the layouters are the same.
+            {
+                //panic!("{:#?}", vk.pinned());
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:678]: as mentioned

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 7]
```diff
@@ -542,34 +649,131 @@ mod tests {
...
+            let test_case_bytes =
+                fs::read("src/utilities/circuit_proof_test_case_short_range_check.bin").unwrap();
+            read_all_proofs(&test_case_bytes[..], 1888).expect("proof must be valid")
+        };
+
+        // Setup phase: generate parameters
+        let params: Params<Affine> = Params::new(11);
+
         // Edge case: zero bits
+        // case 0
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:663]: `// Edge case: zero bits (case 0)`

(no new line)

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 8]
```diff
@@ -542,34 +649,131 @@ mod tests {
...
+                    format!("{:#?}\n", vk.pinned()),
+                    include_str!("vk_short_range_check_0").replace("\r\n", "\n")
+                );
+            }
+
+            // serialized_proof_test_case
+            {
+                match proofs.get(0) {
+                    Some(proof) => {
+                        // Verify the old proofs[0] with the new vk
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:689]: `// Verify the stored proof (case 0) against the generated vk` (all occurrences)

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 9]
```diff
@@ -542,34 +649,131 @@ mod tests {
...
+        // If the environment variable CIRCUIT_TEST_GENERATE_NEW_PROOF is set,
+        // write the old proofs in a file
+        if std::env::var_os("CIRCUIT_TEST_GENERATE_NEW_PROOF").is_some() {
+            let create_proof = || -> std::io::Result<()> {
+                let file = std::fs::File::create(
+                    "src/utilities/circuit_proof_test_case_short_range_check.bin",
+                )?;
+                write_all_test_case(file, &proofs)
+            };
+            create_proof().expect("should be able to write new proof");
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:776]: as mentioned x3

File: halo2_gadgets/src/utilities/test_circuit.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        plonk::verify_proof(params, vk, strategy, &[&[]], &mut transcript)
+    }
+    /// Constructs a new Proof value.
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Proof(bytes)
+    }
+}
+
+/// write proof to a file
+#[allow(dead_code)]
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/test_circuit.rs:58]: remove `#[..]`

Comment (PaulLaux) [reply halo2_gadgets/src/utilities/test_circuit.rs:58]: all locations

File: halo2_gadgets/src/vk_ecc_chip [hunk 11]
```diff
@@ -0,0 +1,9915 @@
+PinnedVerificationKey {
```
Comment (PaulLaux) [inline halo2_gadgets/src/vk_ecc_chip:1]: follow the same file name convention to indicate this is a test case:
`vk_ecc_chip_0`
Same for sinsemilla

File: halo2_gadgets/src/sinsemilla/vk_merkle_chip [hunk 12]
```diff
@@ -0,0 +1,3271 @@
+PinnedVerificationKey {
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/vk_merkle_chip:1]: name `vk_merkle_chip_0`

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 13]
```diff
@@ -279,10 +297,12 @@ pub mod tests {
...
             // Load generator table (shared across both configs)
-            SinsemillaChip::<TestHashDomain, TestCommitDomain, TestFixedBases>::load(
-                config.0.sinsemilla_config.clone(),
-                &mut layouter,
-            )?;
+            SinsemillaChip::<
+                TestHashDomain,
+                TestCommitDomain,
+                TestFixedBases,
+                LookupRangeCheckConfig<pallas::Base, { crate::sinsemilla::primitives::K }>,
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/merkle.rs:304]: why not use `Lookup: DefaultLookupRangeCheck` ? x3

File: halo2_gadgets/src/utilities/lookup_range_check.rs [hunk 14]
```diff
@@ -382,9 +431,18 @@ impl<F: PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K> {
     }
 }
 
+/// The `DefaultLookupRangeCheck` trait extends the `LookupRangeCheck` with additional
+/// standard traits necessary for effective use in cryptographic contexts.
+pub trait DefaultLookupRangeCheck:
```
Comment (PaulLaux) [inline halo2_gadgets/src/utilities/lookup_range_check.rs:436]: Let's call this `PallasLookupRC`

File: halo2_gadgets/src/sinsemilla/merkle.rs [hunk 15]
```diff
@@ -369,17 +387,75 @@ pub mod tests {
...
+        // Read the old proof into 'proof'
+        let proof = {
+            let test_case_bytes =
+                fs::read("src/sinsemilla/circuit_proof_test_case_merkle.bin").unwrap();
+            read_test_case(&test_case_bytes[..]).expect("proof must be valid")
+        };
+
+        // Verify the old proof with the new vk
+        assert!(proof.verify(&vk, &params).is_ok());
+    }
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/merkle.rs:457]: as mentioned

File: halo2_gadgets/src/sinsemilla.rs [hunk 16]
```diff
@@ -738,6 +759,58 @@ pub(crate) mod tests {
...
+        // Read the old proof into 'proof'
+        let proof = {
+            let test_case_bytes = fs::read("src/circuit_proof_test_case_sinsemilla.bin").unwrap();
+            read_test_case(&test_case_bytes[..]).expect("proof must be valid")
+        };
+
+        // Verify the old proof with the new vk
+        assert!(proof.verify(&vk, &params).is_ok());
+    }
+
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla.rs:813]: as mentioned

**Review by PaulLaux — COMMENTED** [review]
good overall, some more polish is required

File: halo2_gadgets/src/sinsemilla/merkle/chip.rs [hunk 17]
```diff
@@ -191,18 +195,67 @@ where
...
         MerkleChip { config }
     }
 }
 
-impl<Hash, Commit, F, const MERKLE_DEPTH: usize>
-    MerkleInstructions<pallas::Affine, MERKLE_DEPTH, { sinsemilla::K }, { sinsemilla::C }>
-    for MerkleChip<Hash, Commit, F>
+/// The MerkleSinsemillaInstructions trait extends the capabilities of SinsemillaInstructions,
+/// for specific cryptographic operations involving generic lookup.
+pub(crate) trait MerkleSinsemillaInstructions<Hash, Commit, F, Lookup>
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/merkle/chip.rs:205]:
```
/// Defines the concrete types of `SinsemillaInstructions` required by the `MerkleCRH`.

File: halo2_gadgets/src/sinsemilla/chip.rs [hunk 18]
```diff
@@ -186,13 +217,15 @@ where
...
+        config
+    }
 
-        // Closures for expressions that are derived multiple times
-        // x_r = lambda_1^2 - x_a - x_p
-        let x_r = |meta: &mut VirtualCells<pallas::Base>, rotation| {
-            config.double_and_add.x_r(meta, rotation)
-        };
+    #[allow(non_snake_case)]
+    fn create_initial_y_q_gate(
```
Comment (PaulLaux) [inline halo2_gadgets/src/sinsemilla/chip.rs:224]: this need documentation

**Review by PaulLaux — COMMENTED** [review]
Added some comments, mostly about style.

In addition,
- `.vk.txt` does not work, let's prefix with: `vk_short_range_check.rdata`
- `.proof.bin` does not work, let's prefix `proof_short_range_check_1.bin`. Also, all proof files should have an index since they are not canonical (unlike vk)

**Comment by ConstanceBeguier** [comment]
Replaced by  #32
